### PR TITLE
Allow fetching arbitrary tag=suffix pairs from GCP/EC2 meta-data

### PR DESCRIPTION
**Pull request by plasticine**
This patch enhances the integration of the agent with EC2 and GCP meta-data endpoints by allowing for arbitrary fetching of tag name & meta-data path suffix pairs.

This new functionality is intended to be additive only and can be used via two new flags, on for AWS and one for GCP; `--tags-from-ec2-meta-data-paths`, `--tags-from-gcp-meta-data-paths`.

#### Example

```
--tags-from-ec2-meta-data-paths "omg=spot/termination-time"
```

This patch also deprecates the `--tags-from-ec2` & `--tags-from-gcp` flags in favour of `--tags-from-ec2-meta-data` && `--tags-from-gcp-meta-data` respectively for consistency. The deprecated flags are not removed yet, but would be in v4.

### CODE REVIEW

File: agent/ec2_meta_data_test.go [hunk 1]
```diff
@@ -0,0 +1,28 @@
+package agent
```
Comment (plasticine) [inline agent/ec2_meta_data_test.go]: Open to ideas for how to test this — ping @lox

File: agent/tags.go [hunk 2]
```diff
@@ -143,3 +205,38 @@ func FetchTags(l *logger.Logger, conf FetchTagsConfig) []string {
 
 	return tags
 }
+
+// TODO: Seems maybe el crapola, dunno will ask lox
```
Comment (plasticine) [inline agent/tags.go]: Well, is it? 😅

File: agent/ec2_meta_data_test.go (hunk 1 above)
Comment (lox) [reply agent/ec2_meta_data_test.go]: I'll 🧠 today.

Comment (plasticine) [reply agent/ec2_meta_data_test.go]: @lox I’m sure it’s pretty easy with a little test http server but I ran out of gojuice to work it out

File: agent/tags.go [hunk 3]
```diff
@@ -95,9 +108,28 @@ func (t *tagFetcher) Fetch(l logger.Logger, conf FetchTagsConfig) []string {
 		}
 	}
 
+	// Attempt to add the EC2 meta-data fetched from user-default path suffixes
+	if len(conf.TagsFromEC2MetaDataPaths) > 0 {
+		paths, err := parseTagValuePathPairs(conf.TagsFromEC2MetaDataPaths)
+		if err != nil {
+			l.Error(fmt.Sprintf("Error parsing meta-data tag and path pairs: %s", err.Error()))
```
Comment (lox) [inline agent/tags.go]: This will show an error but continue, is that the intention?

File: agent/tags.go [hunk 4]
```diff
@@ -123,9 +155,27 @@ func (t *tagFetcher) Fetch(l logger.Logger, conf FetchTagsConfig) []string {
...
+		}
+	}
+
 	// Attempt to add the Google Cloud meta-data
-	if conf.TagsFromGCP {
-		gcpTags, err := t.gcpMetadata()
+	if len(conf.TagsFromGCPMetaDataPaths) > 0 {
+		paths, err := parseTagValuePathPairs(conf.TagsFromGCPMetaDataPaths)
+		if err != nil {
+			l.Error(fmt.Sprintf("Error parsing meta-data tag and path pairs: %s", err.Error()))
```
Comment (lox) [inline agent/tags.go]: Same as the other one.

File: agent/tags.go (hunk 2 above)
Comment (lox) [reply agent/tags.go]: You could probably just use `strings.SplitN` to force it to split into two chunks and then check the len is 2.

File: agent/tags.go [hunk 5]
```diff
@@ -164,3 +214,29 @@ func (t *tagFetcher) Fetch(l logger.Logger, conf FetchTagsConfig) []string {
...
+	for _, pair := range paths {
+		// Sanity check the format of each pair to ensure that it’s parseable
+		index := strings.LastIndex(pair, "=")
+		if index == -1 || index == 0 || index == len(pair)-1 {
+			return result, errors.New(fmt.Sprintf("`%s` cannot be parsed, format should be `tag=metadata/path`", pair))
+		}
+
+		x := strings.Split(pair, "=")
+		// TODO: Should we just let people have stupid keys? Probably?
+		key := strings.ToLower(strings.Trim(x[0], " "))
```
Comment (lox) [inline agent/tags.go]: strings.TrimSpace ftw

**Comment by lox** [comment]
This is looking good @plasticine. Mind just grepping for the config keys that you've deprecated and make sure the configs we ship for packages are updated?

File: agent/tags.go (hunk 3 above)
Comment (plasticine) [reply agent/tags.go]: Good catch — nope, this should ‼️💥

Comment (plasticine) [reply agent/tags.go]: Actually — now I think about it a little more I’m actually not sure about this. I think it’s fine to throw an error but continue and it’s how it already works...

**Pull request merged**